### ISSUE TITLE
Missing setFoo() in ImmutableType.Builder when foo() property is an ImmutableList<T>

### ISSUE THREAD

**Issue by bgertzfield** [immutables/immutables#91]
In Immutables.org 1.1.2, we worked around `List<T>` properties not having a `setFoo()` by using `ImmutableList<T>`.

It looks like the latest master (237dbfa603f90e18c50e710c1e1a56127a5e137c) no longer emits `setFoo()` for `ImmutableList<T>`.

It'd be nice if both `List<T>` and `ImmutableList<T>` had `setFoo()`, but if this is intended to be a breaking change, we can of course work around it with `addAllFoo()`. (We don't actually need to replace the contents of the list.)

**Comment by bgertzfield** [comment]
The same thing applies for other `ImmutableX` Guava types (`ImmutableSet`, `ImmutableMap`, etc.)

This breaks a bunch of places in Buck when we try to upgrade from 1.1.2 to 2.0-SNAPSHOT.

**Comment by elucash** [comment]
We had couple of users (I don't know if I should include Buck team) who prefer to use `ImmutableList` etc as attribute type, and still want special support for collections. 2.0 version recognizes collection interfaces as well as immutable classes for, but other implementations like ArrayList will not be specially supported.

If you decide to migrate, following workarounds may be applicable

```java
// 1. replace setFoo with putAll. this is most straight forward.
// Downside - collection will be rebuilt
builder
    .putAllFoo(immutableCollection)
    .build();

// 2. use withFoo after instance is constructed - this way collection will not be rebuilt.
// Downside - you cannot use it if you validate collection content in `@Check` method
builder
    .build()
    .withFoo(immutableCollection);

// 3. turn collection into nothing special
// currently it could be done using @Default or @Nullable, both are not universally applicable for this
@Value.Default
public  ImmutableList<String> getFoo() {
   return ImmutableList.of();
}

builder
    .setFoo(immutableCollection)
    .build();
```

Have any ideas? If it's ok from the API standpoint in a long run?
Should we put this new behavior under style annotation attribute?

Also there's no rush to move to 2.0, which is not released. And 1.1.x will be supported for some time, could be patched with critical fixes.

**Comment by bgertzfield** [comment]
The first workaround is fine, I just wanted to make sure you knew it was a breaking change. It'll probably affect a lot of people, so maybe the first release should make the new behavior opt-in?

**Comment by elucash** [comment]
That's the dilemma, with 2.0 we want to leave core API mostly unchanged (both annotation and generated), but we do want to fix a lot of inconsistencies, get rid of unwanted dependencies etc.
We expect to properly document breaking changes for 2.0, but to decide on things like opt-in vs opt-out sometimes there's not enough info. The balance between "configurability" vs "pushing all design decisions on users" is very difficult to get right for us.

I got an impression from other users that they mostly use `List` and some of them would use `ImmutableList` instead of a plain list if Immutables would generate `add` and `addAll` initializers on builder. Is behavior in 1.1.x valuable for compatibility or is it legit and convenient on it's own?
I'm interested, because I think we might miss some important aspect / use case here, and it's not too late to change/opt-in/out-out before 2.0 will be released.

**Comment by bgertzfield** [comment]
If it's possible to support `add`, `addAll`, and `set` together for both `List` and `ImmutableList` etc., then could we possibly do that? Then everyone is happy.

**Comment by elucash** [comment]
That is actually a very good idea! Of course it require some rethinking (related to why it hasn't been done before).  It could be paired with the concept of "strict", forward-only builders: When you opt-in for strict builders you won't have setters for collection generated, and setters for regular attributes could be called only once.

**Comment by bgertzfield** [comment]
Awesome! Opt-in strict behavior makes sense to me.

**Comment by elucash** [comment]
Released in 2.0
